### PR TITLE
Sync fix bookmarks object id duplication

### PR DESCRIPTION
**Pull request by AlexeyBarabash**
Resolves  https://github.com/brave/brave-browser/issues/8325

To fix the profile which have duplicated object_ids and keeping crashing when enabling sync back is a separate issue: https://github.com/brave/brave-browser/issues/8358.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
See issue https://github.com/brave/brave-browser/issues/8325

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_sync/brave_profile_sync_service_impl.cc [hunk 1]
```diff
@@ -523,6 +523,16 @@ void BraveProfileSyncServiceImpl::OnSyncReadyBookmarksModelLoaded() {
...
 }
 
+// static
+void BraveProfileSyncServiceImpl::AddNonClonedBookmarkKeys(
+      BookmarkModel* model) {
+  DCHECK(model);
+  DCHECK(model->loaded());
+  model->AddNonClonedKey("object_id");
+  model->AddNonClonedKey("order");
+  model->AddNonClonedKey("sync_timestamp");
```
Comment (darkdh) [inline components/brave_sync/brave_profile_sync_service_impl.cc]: we also need `version`

File: components/brave_sync/brave_profile_sync_service_impl.h [hunk 2]
```diff
@@ -58,6 +58,7 @@ FORWARD_DECLARE_TEST(BraveSyncServiceTest, ProcessOtherBookmarksFolder);
 FORWARD_DECLARE_TEST(BraveSyncServiceTest, ProcessOtherBookmarksChildren);
 FORWARD_DECLARE_TEST(BraveSyncServiceTest, CheckOtherBookmarkRecord);
 FORWARD_DECLARE_TEST(BraveSyncServiceTest, CheckOtherBookmarkChildRecord);
+FORWARD_DECLARE_TEST(BraveSyncServiceTest, AddNonClonedBookmarkKeys);
```
Comment (darkdh) [inline components/brave_sync/brave_profile_sync_service_impl.h]: Why do we need this?  the test only accesses `AddNonClonedBookmarkKeys` which is a public static method

File: components/brave_sync/brave_profile_sync_service_impl.h [hunk 3]
```diff
@@ -231,6 +234,7 @@ class BraveProfileSyncServiceImpl
                            CheckOtherBookmarkRecord);
   FRIEND_TEST_ALL_PREFIXES(::BraveSyncServiceTest,
                            CheckOtherBookmarkChildRecord);
+  FRIEND_TEST_ALL_PREFIXES(::BraveSyncServiceTest, AddNonClonedBookmarkKeys);
```
Comment (darkdh) [inline components/brave_sync/brave_profile_sync_service_impl.h]: ditto

**Comment by AlexeyBarabash** [comment]
CI build on Windows failed with error
```
21:42:06  ninja: build stopped: subcommand failed.
21:42:06  null
21:42:06  powershell.exe : null
21:42:06  At C:\_dup@tmp\durable-ddf68735\powershellWrapper.ps1:3 char:1
21:42:06  + & powershell -NoProfile -NonInteractive -ExecutionPolicy Bypass -Comm ...
21:42:06  + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
21:42:06      + CategoryInfo          : NotSpecified: (null:String) [], RemoteException
21:42:06      + FullyQualifiedErrorId : NativeCommandError
```
restarted

**Comment by AlexeyBarabash** [comment]
now failed iOS build only with 
```
15:50:58  ERROR at //build/timestamp.gni:31:19: Script returned non-zero exit code.
15:50:58  build_timestamp = exec_script(compute_build_timestamp,
15:50:58                    ^----------
15:50:58  Current dir: /Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/out/ios_Release/
15:50:58  Command: python /Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/build/compute_build_timestamp.py official
15:50:58  Returned 1.
```
Restarted.

**Comment by AlexeyBarabash** [comment]
This time again succeded all except iOS build: https://ci.brave.com/job/brave-browser-build-pr/job/sync_fix_obj_id_dup/3/flowGraphTable/ .
```
20:56:20  ERROR at //build/timestamp.gni:31:19: Script returned non-zero exit code.
20:56:20  build_timestamp = exec_script(compute_build_timestamp,
20:56:20                    ^----------
20:56:20  Current dir: /Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/out/ios_Release/
20:56:20  Command: python /Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/build/compute_build_timestamp.py official
20:56:20  Returned 1.
20:56:20  stderr:
```
So re-launch build ignoring all platforms except iOS.

**Comment by AlexeyBarabash** [comment]
iOS build still fails
```
cc.py\" "
05:20:51  ERROR at //build/timestamp.gni:31:19: Script returned non-zero exit code.
05:20:51  build_timestamp = exec_script(compute_build_timestamp,
05:20:51                    ^----------
05:20:51  Current dir: /Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/out/ios_Release/
05:20:51  Command: python /Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/build/compute_build_timestamp.py official
05:20:51  Returned 1.
05:20:51  stderr:
05:20:51  
05:20:51  Traceback (most recent call last):
05:20:51    File "/Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/build/compute_build_timestamp.py", line 127, in <module>
05:20:51      sys.exit(main())
05:20:51    File "/Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/build/compute_build_timestamp.py", line 113, in main
05:20:51      last_commit_timestamp = int(open(lastchange_file).read())
05:20:51  IOError: [Errno 2] No such file or directory: '/Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/build/util/LASTCHANGE.committime'
05:20:51  
05:20:51  See //base/BUILD.gn:34:1: whence it was imported.
```
restarting CI

**Comment by AlexeyBarabash** [comment]
got the same
```
15:32:57  /Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src: gn gen /Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/out/ios_Release --args="fieldtrial_testing_like_official_build=true brave_services_key=\"****\" root_extra_deps=[\"//brave\"] is_component_build=false target_cpu=\"x64\" is_official_build=true is_debug=false dcheck_always_on=false brave_channel=\"nightly\" google_api_key=\"****token\" brave_product_name=\"Brave\" brave_project_name=\"brave\" brave_version_major=\"1\" brave_version_minor=\"7\" brave_version_build=\"10\" chrome_version_string=\"80.0.3987.116\" chrome_version_major=\"80\" webcompat_report_api_endpoint=\"https://webcompat.brave.com/1/webcompat\" enable_cdm_host_verification=false skip_signing=true target_os=\"ios\" enable_dsyms=false enable_stripping=false use_xcode_clang=true use_clang_coverage=false ios_deployment_target=\"12.0\" ios_enable_code_signing=false cc_wrapper=\"/Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/brave/script/redirect-cc.py\" "
15:32:58  ERROR at //build/timestamp.gni:31:19: Script returned non-zero exit code.
15:32:58  build_timestamp = exec_script(compute_build_timestamp,
15:32:58                    ^----------
15:32:58  Current dir: /Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/out/ios_Release/
15:32:58  Command: python /Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/build/compute_build_timestamp.py official
15:32:58  Returned 1.
15:32:58  stderr:
15:32:58  
15:32:58  Traceback (most recent call last):
15:32:58    File "/Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/build/compute_build_timestamp.py", line 127, in <module>
15:32:58      sys.exit(main())
15:32:58    File "/Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/build/compute_build_timestamp.py", line 113, in main
15:32:58      last_commit_timestamp = int(open(lastchange_file).read())
15:32:58  IOError: [Errno 2] No such file or directory: '/Users/jenkins/jenkins/workspace/ser-build-pr_sync_fix_obj_id_dup/src/build/util/LASTCHANGE.committime'
15:32:58  
15:32:58  See //base/BUILD.gn:34:1: whence it was imported.
15:32:58  import("//build/timestamp.gni")
15:32:58  ^-----------------------------
15:32:58  See //BUILD.gn:70:5: which caused the file to be included.
15:32:58      "//base:base_perftests",
15:32:58      ^----------------------
15:32:58  null
15:32:58  null
```

**Comment by btlechowski** [comment]
LGTM

Verification passed on

Brave | 1.7.35 Chromium: 80.0.3987.122 (Official Build) nightly (64-bit)
-- | --
Revision | cf72c4c4f7db75bc3da689cd76513962d31c7b52-refs/branch-heads/3987@{#943}
OS | Ubuntu 18.04 LTS

Brave | 1.7.35 Chromium: 80.0.3987.122 (Official Build) nightly (64-bit)
-- | --
Revision | cf72c4c4f7db75bc3da689cd76513962d31c7b52-refs/branch-heads/3987@{#943}
OS | Windows 7 Service Pack 1 (Build 7601.24544)

Reproduced the crash on 1.4.95

Verified the test plan from https://github.com/brave/brave-browser/issues/8325#issue-567786768, no crash
Verified no crash for bookmark folders.
Verified that only one bookmark(original bookmark) is being initially synced out of all bookmarks with the same id.
Verified that all bookmarks are eventually synced 
Verified no crash with multiple bookmarks with the same id
Verified renaming the bookmarks works
Verified moving of the bookmarks works
Verified additional bookmarks are synced

**Pull request merged**